### ISSUE TITLE
AMP Review 1

### ISSUE THREAD

**Issue by andrewrreed** [fastforwardlabs/few-shot-text-classification#31]
# AMP Review 1

This is one good looking code repository! After reviewing this, my eyes are open to how much room I have for improving quality of my code. Also, because everything is so organized, clean, and well documented, it makes the review process very straightforward... so thanks!

Overall the content is great and all the _major_ things work. I've noted a few suggestions in documentation + some trouble I ran into when running some notebooks and scripts below:

## General Feedback

- [x] When I first tried to launch the project as an AMP, the project failed to create because the default branch was named `main` rather than `master`... looks like [this issue](https://jira.cloudera.com/browse/DSE-12921) has been resolved but the CML version I am testing on doesn't include the bug fix...So, i've temporarily renamed the `main` branch to `master` for testing purposes, just an FYI.
- [x] It looks like `scripts/visualize_text_examples.py` is empty and probably in here by mistake? If so, can remove..

## README.md

- [x] Spelling correction
![image](https://user-images.githubusercontent.com/23462538/106653186-76406a00-6564-11eb-80cd-3cc9af9cc90b.png)
- [x] Can we include a link to the AG News Dataset here:
![image](https://user-images.githubusercontent.com/23462538/106653455-d0412f80-6564-11eb-83b0-e46da6867fac.png)
- [x] I don't think the README currently mentions the infrastructure requirements (Session/Job) for running the scripts/notebooks... I think 4vCPU / 8GB RAM should be good, but might want to add a comment that says so in the readme

### few-shot_text_classification.py

- [x] Training with `device == 'cpu'` currently fails. Looks like you have a comment in `models/few_shot.py` that says "For GPU" - so cpu version isn't implemented? If its a ton of work to implement, I'm ok with just specifying in all the documentation that the training scripts (and notebooks?) can only be run with a GPU. Otherwise, we should probably add in the functionality for CPU so users don't get frustrated.
![image](https://user-images.githubusercontent.com/23462538/106659398-ac81e780-656c-11eb-8018-92f3041b98ab.png)

### CreatingRedditDataset.ipynb

- [x] Subsampled dataset name is incorrect - should be `filename = "../data/reddit/reddit_dataset_1300.pt"`
- [x] Actually, now that I look at this, I don't think this notebook is meant to be run...right? If this is here to provide transparency for how the dataset was created, but not intended to be runnable b/c data size, thats fine. I'd say let's just make it very clear that this notebook isn't intended to be run.

### Wmap_Experiments.ipynb

- [x] These setup instructions only apply if not launching as an AMP - would probably be good to comment them out and add a comment / blurb saying "If you did not launch as an AMP via CML, run these cells first..."
![image](https://user-images.githubusercontent.com/23462538/106661359-1b604000-656f-11eb-8739-c6cbf5d51f73.png)

- [x] I think there has been some refactoring since this notebook was written so the notebook doesn't run in its current state:
![image](https://user-images.githubusercontent.com/23462538/106658272-139e9c80-656b-11eb-88ff-3074873c23ca.png)

### Zmap_Experiments.ipynb

- [x] Same as Wmap_Experiments comments above

## Few-Shot UI Application

- [x] I'm not entirely sure why, and there doesn't seem to be any rhyme or reason to when this happens, but if I toggle around with different example/classifier enhancement combo's, I _occasionally_ am able to break the UMAP scatter plot. This might be a performance induced bug because its not repeatable, nor consistent... have you encountered this at all?
![image](https://user-images.githubusercontent.com/23462538/106655134-fa93ec80-6566-11eb-8a9f-73584595bec4.png)

**Comment by melaniebeck** [comment]
Thanks for the thoughtful feedback, Andrew. I believe I've addressed the bulk of your comments.

Regarding your last point about the Streamlit app, I have not yet come across that issue but I have noticed other buggy things: namely, when a user changes the list of possible labels, the bar chart below somehow remembers the old set of labels and I have *no* idea how to make it stop doing that. I suspect this issue is due to the way Streamlit caches things. Perhaps the image issue you pointed out is similar.  If so, I'm honestly not sure how to make them more robust.

**Comment by andrewrreed** [comment]
### General

Overall, this looks great all the core pieces are clean and functional. A few general checklist items, followed by some thoughts about the notebooks:

- [x] Let's not worry about the Streamlit bugs because, well, they're Streamlit bugs
- [x] Are you ok with this repo name for the public Cloudera org (if not, feel free to update to whatever you like): **_CML_AMP_Few-Shot_Text_Classification_**
- [x] Was there any borrowed code that needs to be referenced for this repo?
- [x] Wanna pick a cover image for the AMP repo? Or I can just snag the one from the README?

### Notebooks

Copying your note from Slack for context:

> Some important notes: the notebooks are pretty much for show, not for execution. I’ve made this a lot more clear now. Technically many of the cells can be run but probably shouldn’t be without a GPU. They would just take way too long. On the other hand, many cells can’t be run because I don’t include all the data resources required for them to execute properly.  Maybe this means we don’t include them in the AMP? But I feel they do need to stay with this particular repo because they are designed to mirror the report. 

I agree that these notebooks should stay with the repo because they complement the report, and I think its perfectly ok that they don't execute in full for the reasons listed above. With that I do have a couple of thoughts to make this a bit clearer to a user who is first looking at the repo:

- [x] Let's update the language in the README so that new users have a clear understanding and expectation of what the notebooks are/aren't upfront. In particular, removing the word "interactive" and stating that they are not intended to be run in full because of missing/too large data that isn't available.
![image](https://user-images.githubusercontent.com/23462538/108080184-96305d00-703d-11eb-9a37-7e326754e769.png)

- [x] Since, we are recommending Colab as the place to view and interact with the notebooks, maybe let's just include the direct link so users can very easily open them up. So place, [this link](https://colab.research.google.com/github/fastforwardlabs/few-shot-text-classification/blob/master/notebooks/Wmap_Experiments.ipynb) in the header section of each notebook?
![image](https://user-images.githubusercontent.com/23462538/108081199-aac12500-703e-11eb-8bde-9882dc409859.png)

- [x] Last thing on notebooks... I promise.... When I opened each notebook to run in Colab, each one had an issue importing the `fewshot` library upfront. I think if we are going to tell users that these notebooks mostly work, except for missing/unusable dataset problems, then the general setup/import steps _should_ be functional. However, if we just want to make a blanket statement that the notebooks _are not_ useable and should not be run at all, then we can leave as is, but update the verbiage to match.


#### Zmap_Experiments.ipynb
![image](https://user-images.githubusercontent.com/23462538/108079020-56b54100-703c-11eb-9ba3-9ff9f4805f77.png)

#### Wmap_Experiments.ipynb
![image](https://user-images.githubusercontent.com/23462538/108079453-d216f280-703c-11eb-8fa2-994a8e699c28.png)

**Comment by melaniebeck** [comment]
Thanks again Andrew for the additional feedback - these notebooks are going to be amazing (and I've been reconsidering whether I could include all of the necessary data products in order to get them to execute top to bottom -- for now I'm leaving it as is but v.2.0 will likely include more data). I've also updated the README as you suggested -- and good call on that!

As to your first few questions/comments:
*  Repo name looks fine to me! 
* While the overall idea for this project came from a HF blog post, they didn't offer much in the way of code examples for this particular technique. I had to roll my own for pretty much everything. One thing ended up being the same, a 1-liner call to a pytorch cosine similarity function because there's no other way to call it -- it's straight from the Pytorch documentation. For that reason, I didn't link directly to it. If you think it would be better to err the other way I can go ahead and create a code doc.
* Go ahead and snag the repo image! Thanks!
